{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "f2159562",
   "metadata": {},
   "source": [
    "# iMARGI configuration file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "f2dc73f1",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2024-10-15T09:02:50.313977Z",
     "iopub.status.busy": "2024-10-15T09:02:50.313840Z",
     "iopub.status.idle": "2024-10-15T09:02:50.320507Z",
     "shell.execute_reply": "2024-10-15T09:02:50.319980Z"
    }
   },
   "outputs": [],
   "source": [
    "imargi = '''\n",
    "\n",
    "params {\n",
    "    // INPUT OPTIONS:\n",
    "    input                       = null\n",
    "    exp_type                    = 'imargi'\n",
    "    procedure                   = 'old'\n",
    "    split_by_chromosomes        = true\n",
    "    \n",
    "    // PROCESSING TOOLS:       --------------------------------------------------------------------------------\n",
    "    dedup_tool                  = \"fastq-dupaway\"    // options: \"fastq-dupaway\", \"fastuniq\", \"climpify\"\n",
    "    trim_tool                   = \"fastp\"            // options: \"trimmomatic\", \"bbduk\", \"cutadapt\", \"fastp\"\n",
    "    align_tool                  = 'bwa_mem_hisat'           // options: \"hisat2\", \"bowtie2\", \"star\", \"bwa\"\n",
    "\n",
    "    // REFERENCE:              --------------------------------------------------------------------------------\n",
    "    genome                      = 'GRCh38'\n",
    "    genome_fasta                = '/nfs/ilnitsky/nf-rnachrom/reference/hs/hg38/GRCh38.p13.fa'\n",
    "    hisat2_index                = '/nfs/ilnitsky/nf-rnachrom/reference/hs/hg38'\n",
    "    bwa_index                   = '/nfs/ilnitsky/nf-rnachrom/reference/hs/bwa'\n",
    "    star_index                  = '/nfs/ilnitsky/nf-rnachrom/reference/hs/STAR'\n",
    "    bowtie2_index \t\t\t\t= '/nfs/imarkov/arabidopsis/rnachrom_launch/genome'\n",
    "    splice_sites                = null\n",
    "    stages = 'annotation, splicing, normalize, peak_calling'\n",
    "\n",
    "    // ANNOTATION:             --------------------------------------------------------------------------------\n",
    "    annot_BED                   = \"/nfs/ilnitsky/nf-rnachrom/reference/hs/gencode.v24.annotation.uniq.bed\"\n",
    "    annot_GTF                   = \"/nfs/ilnitsky/nf-rnachrom/reference/hs/gencode.v24.annotation.gtf\"\n",
    "    blacklist                   = '/nfs/ilnitsky/nf-rnachrom/reference/hs/hg38.blacklist.bed'\n",
    "    chromsizes                  = \"/nfs/ilnitsky/nf-rnachrom/reference/hs/hg38_canonical_chromsizes.tsv\"\n",
    "    detect_strand_genes_list    = \"${projectDir}/assets/gencode_v43_rpl_genes.txt\"       \n",
    "\n",
    "    // RESTRICTION SITES & BRIDGE SEARCH:  --------------------------------------------------------------------\n",
    "    dna_part_processing         = \"s[CT|AGCT]*\"\n",
    "    rna_part_processing         = \"-[2].\"\n",
    "    bridge_processing           = false\n",
    "    description_sequence        = params.bridge_processing ? \"${dna_part_processing}b${forward_bridge_seq}(${max_mismatches})${rna_part_processing}\" : \"\"\n",
    "    //                         --------------------------------------------------------------------------------\n",
    "\n",
    "    // OTHER OPTIONS           --------------------------------------------------------------------------------\n",
    "    smartseq_filter             = false\n",
    "    // Max resource options\n",
    "    max_memory                 = '128.GB'\n",
    "    max_cpus                   = 16\n",
    "    max_time                   = '240.h'\n",
    "    // Boilerplate options\n",
    "    outdir                     = null\n",
    "    publish_dir_mode           = 'copy'\n",
    "    email                      = null\n",
    "    help                       = false\n",
    "    version                    = false\n",
    "    //                         --------------------------------------------------------------------------------\n",
    "    \n",
    "    // BARDIC OPTIONS          --------------------------------------------------------------------------------\n",
    "\n",
    "    //                         --------------------------------------------------------------------------------\n",
    "}\n",
    "\n",
    "conda.cacheDir = \"/home/ilnitsky/nf-rnachrom/conda_env\"\n",
    "process {\n",
    "   withName: '.*' {\n",
    "       cpus = 30\n",
    "       memory = 60.GB\n",
    "   }\n",
    "}\n",
    "\n",
    "flags {\n",
    "    // COMMAND FLAGS  --------------------------------------------------------------------------------\n",
    "    fastq_dupaway           = \"--format fastq --compare-seq loose\"\n",
    "    trimmomatic             = \"SLIDINGWINDOW:5:26 MINLEN:12\"\n",
    "    fastp                   = \"-5 --correction --cut_window_size 5 --cut_mean_quality 26\"\n",
    "    pear                    = \"-p 0.01 -v 20 -n 50\"\n",
    "    star                    = \"\"\"--outSAMstrandField intronMotif  --outSAMattributes NH HI AS nM ch                                --chimSegmentMin 12 --chimJunctionOverhangMin 12 --chimOutType WithinBAM                                --alignSJDBoverhangMin 999  --alignIntronMax 100  --outSAMtype BAM  SortedByCoordinate\"\"\"\n",
    "    bwa_mem                 = \"-SP5M\"\n",
    "    bam_filter              = \"-bS -F 4 -e '[NH]==1'\"\n",
    "    //               --------------------------------------------------------------------------------\n",
    "}\n",
    "\n",
    "env {\n",
    "    PYTHONNOUSERSITE = 1\n",
    "    PYTHONPATH       = \"/home/ilnitsky/anaconda3/envs/ken/bin/python\"\n",
    "    R_PROFILE_USER   = \"/.Rprofile\"\n",
    "    R_ENVIRON_USER   = \"/.Renviron\"\n",
    "    JULIA_DEPOT_PATH = \"/home/ilnitsky/.julia\"\n",
    "}\n",
    "\n",
    "includeConfig \"${projectDir}/conf/base.config\"\n",
    "includeConfig \"${projectDir}/conf/modules.config\"\n",
    "\n",
    "'''\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "08aa263f",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ed14d69e",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
