

<!DOCTYPE html>
<html class="writer-html5" lang="en" data-content_root="./">
<head>
  <meta charset="utf-8" /><meta name="viewport" content="width=device-width, initial-scale=1" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>RNAchrom Configuration &mdash; nf-rnachrom 1 documentation</title>
      <link rel="stylesheet" type="text/css" href="_static/pygments.css?v=80d5e7a1" />
      <link rel="stylesheet" type="text/css" href="_static/css/theme.css?v=e59714d7" />

  
      <script src="_static/jquery.js?v=5d32c60e"></script>
      <script src="_static/_sphinx_javascript_frameworks_compat.js?v=2cd50e6c"></script>
      <script src="_static/documentation_options.js?v=29a6c3e3"></script>
      <script src="_static/doctools.js?v=9bcbadda"></script>
      <script src="_static/sphinx_highlight.js?v=dc90522c"></script>
      <script crossorigin="anonymous" integrity="sha256-Ae2Vz/4ePdIu6ZyI/5ZGsYnb+m0JlOmKPjt6XZ9JJkA=" src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
    <script src="_static/js/theme.js"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="Stages of Data Analysis" href="stages.html" />
    <link rel="prev" title="Input file preparation" href="input.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >

          
          
          <a href="index.html" class="icon icon-home">
            nf-rnachrom
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <ul>
<li class="toctree-l1"><a class="reference internal" href="index.html">Overview</a></li>
</ul>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="installation.html">Setup and Installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="input.html">Input file preparation</a></li>
<li class="toctree-l1 current"><a class="current reference internal" href="#">RNAchrom Configuration</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#input-options">Input Options</a></li>
<li class="toctree-l2"><a class="reference internal" href="#processing-tools">Processing Tools</a></li>
<li class="toctree-l2"><a class="reference internal" href="#reference-genome">Reference Genome</a></li>
<li class="toctree-l2"><a class="reference internal" href="#annotation-files">Annotation Files</a></li>
<li class="toctree-l2"><a class="reference internal" href="#restriction-sites-bridge-processing">Restriction Sites &amp; Bridge Processing</a></li>
<li class="toctree-l2"><a class="reference internal" href="#other-options">Other Options</a></li>
<li class="toctree-l2"><a class="reference internal" href="#command-flags">Command Flags</a></li>
<li class="toctree-l2"><a class="reference internal" href="#environment-variables">Environment Variables</a></li>
<li class="toctree-l2"><a class="reference internal" href="#included-configs">Included Configs</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="stages.html">Stages of  Data Analysis</a></li>
<li class="toctree-l1"><a class="reference internal" href="results.html">Results</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="index.html">nf-rnachrom</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="index.html" class="icon icon-home" aria-label="Home"></a></li>
      <li class="breadcrumb-item active">RNAchrom Configuration</li>
      <li class="wy-breadcrumbs-aside">
            <a href="_sources/configuration.rst.txt" rel="nofollow"> View page source</a>
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <section id="rnachrom-configuration">
<span id="id1"></span><h1>RNAchrom Configuration<a class="headerlink" href="#rnachrom-configuration" title="Link to this heading"></a></h1>
<p>The RNAchrom workflow uses a configuration file to specify various parameters and options. This document describes the key sections and parameters in the configuration file.
We have provided a list of configurations most suitable for data processing of each experiment type. While core tools and parameters are constant across pipelines to ensure reliability,
each configuration uniquely tailors settings to align with the specific analysis type. Those options are the ones used while preparing raw files
for <a class="reference external" href="https://rnachrom2.bioinf.fbb.msu.ru">RNA-Chrom</a>  database.</p>
<p>The pipeline supports various all-to-all experiment types, each with its own configuration file. Click on the configuration file name to view more details.</p>
<ul class="simple">
<li><p><a class="reference internal" href="examples/iMARGI.html"><span class="doc">iMARGI</span></a>: iMARGI experiment configuration</p></li>
<li><p><a class="reference internal" href="examples/grid.html"><span class="doc">GRID</span></a>: GRID experiment configuration</p></li>
<li><p><a class="reference internal" href="examples/char.html"><span class="doc">CHAR-seq</span></a>: CHAR-seq experiment configuration</p></li>
<li><p><a class="reference internal" href="examples/redc.html"><span class="doc">Red-C</span></a>: Red-C experiment configuration</p></li>
<li><p><a class="reference internal" href="examples/redchip.html"><span class="doc">Red-ChIP</span></a>: Red-ChIP experiment configuration</p></li>
</ul>
<p>These configs share a common structure but differ in key aspects:</p>
<ul class="simple">
<li><p>Experiment type</p></li>
<li><p>Bridge processing settings and sequences</p></li>
<li><p>Alignment tools and parameters</p></li>
<li><p>Other specific flags for processing tools</p></li>
</ul>
<p>Below you can explore options in details.</p>
<section id="input-options">
<h2>Input Options<a class="headerlink" href="#input-options" title="Link to this heading"></a></h2>
<table class="docutils align-default">
<colgroup>
<col style="width: 30.0%" />
<col style="width: 70.0%" />
</colgroup>
<thead>
<tr class="row-odd"><th class="head"><p>Parameter</p></th>
<th class="head"><p>Description</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p><code class="docutils literal notranslate"><span class="pre">input</span></code></p></td>
<td><p>Path to input file(s)</p></td>
</tr>
<tr class="row-odd"><td><p><code class="docutils literal notranslate"><span class="pre">exp_type</span></code></p></td>
<td><p>Type of experiment (e.g. ‘grid’, ‘radicl’, ‘imargi’)</p></td>
</tr>
<tr class="row-even"><td><p><code class="docutils literal notranslate"><span class="pre">procedure</span></code></p></td>
<td><p>Processing procedure (‘old’ or ‘new’)</p></td>
</tr>
<tr class="row-odd"><td><p><code class="docutils literal notranslate"><span class="pre">split_by_chromosomes</span></code></p></td>
<td><p>Whether to split processing by chromosomes (true/false)</p></td>
</tr>
</tbody>
</table>
</section>
<section id="processing-tools">
<h2>Processing Tools<a class="headerlink" href="#processing-tools" title="Link to this heading"></a></h2>
<p>Those options determine which specific tools are used at each step of the workflow, affecting processing speed and output quality.
All available tools from modules are listed below:</p>
<table class="docutils align-default">
<colgroup>
<col style="width: 30.0%" />
<col style="width: 70.0%" />
</colgroup>
<thead>
<tr class="row-odd"><th class="head"><p>Parameter</p></th>
<th class="head"><p>Description</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p><code class="docutils literal notranslate"><span class="pre">dedup_tool</span></code></p></td>
<td><p>Tool for deduplication (options: “fastq-dupaway”, “fastuniq”, “climpify”)</p></td>
</tr>
<tr class="row-odd"><td><p><code class="docutils literal notranslate"><span class="pre">trim_tool</span></code></p></td>
<td><p>Tool for read trimming (options: “trimmomatic”, “bbduk”, “cutadapt”, “fastp”)</p></td>
</tr>
<tr class="row-even"><td><p><code class="docutils literal notranslate"><span class="pre">align_tool</span></code></p></td>
<td><p>Tool for alignment (options: “hisat2”, “bowtie2”, “star”)</p></td>
</tr>
<tr class="row-odd"><td><p><code class="docutils literal notranslate"><span class="pre">merge_pairedend_tool</span></code></p></td>
<td><p>Tool for merging paired-end reads (options: “bbmerge”, “pear”)</p></td>
</tr>
</tbody>
</table>
</section>
<section id="reference-genome">
<h2>Reference Genome<a class="headerlink" href="#reference-genome" title="Link to this heading"></a></h2>
<table class="docutils align-default">
<colgroup>
<col style="width: 30.0%" />
<col style="width: 70.0%" />
</colgroup>
<thead>
<tr class="row-odd"><th class="head"><p>Parameter</p></th>
<th class="head"><p>Description</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p><code class="docutils literal notranslate"><span class="pre">genome</span></code></p></td>
<td><p>Reference genome identifier (“GRCh38”, “GRCm38” “TAIR10”,  … )</p></td>
</tr>
<tr class="row-odd"><td><p><code class="docutils literal notranslate"><span class="pre">genome_fasta</span></code></p></td>
<td><p>Path to reference genome FASTA file</p></td>
</tr>
<tr class="row-even"><td><p><code class="docutils literal notranslate"><span class="pre">hisat2_index</span></code>; <code class="docutils literal notranslate"><span class="pre">bwa_index</span></code>; <code class="docutils literal notranslate"><span class="pre">star_index</span></code></p></td>
<td><p>Path to aligner index files</p></td>
</tr>
<tr class="row-odd"><td><p><code class="docutils literal notranslate"><span class="pre">splice_sites</span></code></p></td>
<td><p>Path to splice sites file (optional)</p></td>
</tr>
</tbody>
</table>
</section>
<section id="annotation-files">
<h2>Annotation Files<a class="headerlink" href="#annotation-files" title="Link to this heading"></a></h2>
<table class="docutils align-default">
<colgroup>
<col style="width: 30.0%" />
<col style="width: 70.0%" />
</colgroup>
<thead>
<tr class="row-odd"><th class="head"><p>Parameter</p></th>
<th class="head"><p>Description</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p><code class="docutils literal notranslate"><span class="pre">annot_BED</span></code></p></td>
<td><p>Path to BED annotation file</p></td>
</tr>
<tr class="row-odd"><td><p><code class="docutils literal notranslate"><span class="pre">annot_GTF</span></code></p></td>
<td><p>Path to GTF annotation file</p></td>
</tr>
<tr class="row-even"><td><p><code class="docutils literal notranslate"><span class="pre">blacklist</span></code></p></td>
<td><p>Path to blacklist regions file</p></td>
</tr>
<tr class="row-odd"><td><p><code class="docutils literal notranslate"><span class="pre">chromsizes</span></code></p></td>
<td><p>Path to chromosome sizes file</p></td>
</tr>
<tr class="row-even"><td><p><code class="docutils literal notranslate"><span class="pre">detect_strand_genes_list</span></code></p></td>
<td><p>Path to gene list for strand detection</p></td>
</tr>
</tbody>
</table>
</section>
<section id="restriction-sites-bridge-processing">
<h2>Restriction Sites &amp; Bridge Processing<a class="headerlink" href="#restriction-sites-bridge-processing" title="Link to this heading"></a></h2>
<table class="docutils align-default">
<colgroup>
<col style="width: 30.0%" />
<col style="width: 70.0%" />
</colgroup>
<thead>
<tr class="row-odd"><th class="head"><p>Parameter</p></th>
<th class="head"><p>Description</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p><code class="docutils literal notranslate"><span class="pre">dna_part_processing</span></code></p></td>
<td><p>Pattern for DNA part processing</p></td>
</tr>
<tr class="row-odd"><td><p><code class="docutils literal notranslate"><span class="pre">rna_part_processing</span></code></p></td>
<td><p>Pattern for RNA part processing</p></td>
</tr>
<tr class="row-even"><td><p><code class="docutils literal notranslate"><span class="pre">bridge_processing</span></code></p></td>
<td><p>Enable bridge processing (true/false)</p></td>
</tr>
<tr class="row-odd"><td><p><code class="docutils literal notranslate"><span class="pre">debridge_tool</span></code></p></td>
<td><p>Tool for debridge processing (options: “bitap”, “chartools”)</p></td>
</tr>
<tr class="row-even"><td><p><code class="docutils literal notranslate"><span class="pre">forward_bridge_seq</span></code></p></td>
<td><p>Forward bridge sequence</p></td>
</tr>
<tr class="row-odd"><td><p><code class="docutils literal notranslate"><span class="pre">reverse_bridge_seq</span></code></p></td>
<td><p>Reverse bridge sequence</p></td>
</tr>
<tr class="row-even"><td><p><code class="docutils literal notranslate"><span class="pre">max_mismatches</span></code></p></td>
<td><p>Maximum allowed mismatches</p></td>
</tr>
<tr class="row-odd"><td><p><code class="docutils literal notranslate"><span class="pre">min_rna_dna_parts_length</span></code></p></td>
<td><p>Minimum length of RNA/DNA parts</p></td>
</tr>
<tr class="row-even"><td><p><code class="docutils literal notranslate"><span class="pre">description_sequence</span></code></p></td>
<td><p>Description of sequence processing</p></td>
</tr>
</tbody>
</table>
</section>
<section id="other-options">
<h2>Other Options<a class="headerlink" href="#other-options" title="Link to this heading"></a></h2>
<table class="docutils align-default">
<colgroup>
<col style="width: 30.0%" />
<col style="width: 70.0%" />
</colgroup>
<thead>
<tr class="row-odd"><th class="head"><p>Parameter</p></th>
<th class="head"><p>Description</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p><code class="docutils literal notranslate"><span class="pre">smartseq_filter</span></code></p></td>
<td><p>Enable Smart-seq filtering (true/false)</p></td>
</tr>
<tr class="row-odd"><td><p><code class="docutils literal notranslate"><span class="pre">max_memory</span></code></p></td>
<td><p>Maximum memory allocation</p></td>
</tr>
<tr class="row-even"><td><p><code class="docutils literal notranslate"><span class="pre">max_cpus</span></code></p></td>
<td><p>Maximum CPU cores to use</p></td>
</tr>
<tr class="row-odd"><td><p><code class="docutils literal notranslate"><span class="pre">max_time</span></code></p></td>
<td><p>Maximum run time</p></td>
</tr>
<tr class="row-even"><td><p><code class="docutils literal notranslate"><span class="pre">outdir</span></code></p></td>
<td><p>Output directory path</p></td>
</tr>
<tr class="row-odd"><td><p><code class="docutils literal notranslate"><span class="pre">publish_dir_mode</span></code></p></td>
<td><p>Mode for publishing output files</p></td>
</tr>
<tr class="row-even"><td><p><code class="docutils literal notranslate"><span class="pre">email</span></code></p></td>
<td><p>Email for notifications</p></td>
</tr>
</tbody>
</table>
</section>
<section id="command-flags">
<h2>Command Flags<a class="headerlink" href="#command-flags" title="Link to this heading"></a></h2>
<p>The configuration file includes specific flags for various tools:</p>
<ul class="simple">
<li><p><code class="docutils literal notranslate"><span class="pre">fastq_dupaway</span></code>: Flags for fastq-dupaway</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">trimmomatic</span></code>: Flags for Trimmomatic</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">fastp</span></code>: Flags for fastp</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">pear</span></code>: Flags for PEAR</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">bam_filter</span></code>: Flags for BAM filtering</p></li>
</ul>
</section>
<section id="environment-variables">
<h2>Environment Variables<a class="headerlink" href="#environment-variables" title="Link to this heading"></a></h2>
<p>The config file also sets various environment variables like <code class="docutils literal notranslate"><span class="pre">PYTHONPATH</span></code>, <code class="docutils literal notranslate"><span class="pre">R_PROFILE_USER</span></code>, etc.</p>
</section>
<section id="included-configs">
<h2>Included Configs<a class="headerlink" href="#included-configs" title="Link to this heading"></a></h2>
<p>The main config file includes additional config files:</p>
<ul class="simple">
<li><p><code class="docutils literal notranslate"><span class="pre">base.config</span></code>: Base configuration for all pipelines</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">modules.config</span></code>: DSL2 module specific options</p></li>
</ul>
<p>This configuration setup allows flexible customization of the RNAchrom workflow for different experimental designs and processing requirements.</p>
</section>
</section>


           </div>
          </div>
          <footer><div class="rst-footer-buttons" role="navigation" aria-label="Footer">
        <a href="input.html" class="btn btn-neutral float-left" title="Input file preparation" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
        <a href="stages.html" class="btn btn-neutral float-right" title="Stages of Data Analysis" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2024, Ivan Ilnitskiy.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>